filterExprParser.go

```diff
@@ -0,0 +1,691 @@
+	// >=
+	return feo.OpChars0 != nil && feo.OpChars0.GreaterThan != nil && feo.OpChars1 != nil && feo.OpChars1.Equal != nil
+}
+
+func (feo *FECompareOp) IsLessThan() bool {
+	// <
+	return feo.OpChars0 != nil && feo.OpChars0.LessThan != nil && feo.OpChars1 == nil
+}
+
+func (feo *FECompareOp) IsLessThanOrEqualTo() bool {
+	// <=
+	return feo.OpChars0 != nil && feo.OpChars0.LessThan != nil && feo.OpChars1 != nil && feo.OpChars1.Equal != nil
+}
+
+func (feo *FECompareOp) String() string {
+	if feo.IsEqual() {
+		return "="
+	} else if feo.IsNotEqual() {
+		return "!="
+	} else if feo.IsGreaterThan() {
+		return ">"
+	} else if feo.IsGreaterThanOrEqualTo() {
+		return ">="
+	} else if feo.IsLessThan() {
+		return "<"
+	} else if feo.IsLessThanOrEqualTo() {
+		return "<="
+	}
+	return ""
+}
+
+type FECheckOp struct {
+	Exists  *bool `@"EXISTS" | ( "IS"`
+	Not     *bool `[ @"NOT" ]`
+	Null    *bool `( @"NULL" |`
+	Missing *bool `@"MISSING" ) )`
+}
+
+func (feco *FECheckOp) IsExists() bool {
+	return feco.Exists != nil && *feco.Exists == true
+}
+
+func (feco *FECheckOp) isNot() bool {
+	return feco.Not != nil && *feco.Not == true
+}
+
+func (feco *FECheckOp) IsMissing() bool {
+	return !feco.isNot() && feco.Missing != nil && *feco.Missing == true
+}
+
+func (feco *FECheckOp) IsNotMissing() bool {
+	return !feco.IsMissing()
+}
+
+func (feco *FECheckOp) IsNull() bool {
+	return !feco.isNot() && feco.Null != nil && *feco.Null == true
+}
+
+func (feco *FECheckOp) IsNotNull() bool {
+	return !feco.IsNull()
+}
+
+func (feco *FECheckOp) String() string {
+	if feco.IsExists() {
+		return "EXISTS"
+	} else if feco.IsMissing() {
+		return "IS MISSING"
+	} else if feco.IsNotMissing() {
+		return "IS NOT MISSING"
+	} else if feco.IsNull() {
+		return "IS NULL"
+	} else if feco.IsNotNull() {
+		return "IS NOT NULL"
+	} else {
+		return "?? (FECheckOp)"
+	}
+}
+
+// Technically we could have an slice of arguments, but having OneArg vs NoArg vs TwoArg could
+// allow us to do more strict function check (i.e. certain funcs should only allow one argument, etc, at this level)
+type FEConstFuncExpression struct {
+	ConstFuncNoArg   *FEConstFuncNoArg   `@@ |`
+	ConstFuncOneArg  *FEConstFuncOneArg  `@@ |`
+	ConstFuncTwoArgs *FEConstFuncTwoArgs `@@`
+}
+
+func (f *FEConstFuncExpression) String() string {
+	if f.ConstFuncNoArg != nil {
+		return f.ConstFuncNoArg.String()
+	} else if f.ConstFuncOneArg != nil {
+		return f.ConstFuncOneArg.String()
+	} else if f.ConstFuncTwoArgs != nil {
+		return f.ConstFuncTwoArgs.String()
+	} else {
+		return "?? (FEConstFuncExpression)"
+	}
+}
+
+type FEConstFuncNoArg struct {
+	ConstFuncNoArgName *FEConstFuncNoArgName `( @@ "(" ")" )`
+}
+
+func (f *FEConstFuncNoArg) String() string {
+	if f.ConstFuncNoArgName != nil {
+		return fmt.Sprintf("%v()", f.ConstFuncNoArgName.String())
+	} else {
+		return "?? (FEConstFuncNoArg)"
+	}
+}
+
+type FEConstFuncNoArgName struct {
+	Pi *bool `@"PI" |` // FuncPi
```

Comment:
here you are doing '@PI |' and '@E'. is this equivalent to '@PI' and '| @E', which is the format used in 
https://github.com/alecthomas/participle

Reply:
Either one is fine